File: 2015/python/16.py
# Advent of Code 2015, day 16
# (c) aichingert

S = {"children": 3, "cats": 7, "samoyeds": 2, "pomeranians": 3, "akitas": 0, "vizslas": 0, "goldfish": 5, "trees": 3, "cars": 2, "perfumes": 1}

def part1(aunts):
    ans = ("", 0)

    for aunt in aunts:
        cur = 0
        for k in S.keys():
            if k in aunt[1] and aunt[1][k] == S[k]:
                cur += 1
        if cur > ans[1]:
            ans = (aunt[0], cur)

    return ans[0]

def part2(aunts):
    ans = ("", 0)

    for aunt in aunts:
        cur = 0
        for k in S.keys():
            if k in aunt[1]:
                if k == "cats" or k == "trees":
                    if S[k] <= aunt[1][k]:
                        cur += 1
                elif k == "pomeranians" or k == "goldfish":
                    if S[k] >= aunt[1][k]:
                        cur += 1
                else:
                    if S[k] == aunt[1][k]:
                        cur += 1

        if cur > ans[1]:
            ans = (aunt[0], cur)

    return ans[0]

s = []
for l in open("../input/16").read().split('\n'):
    sue, vls = l.split(": ", 1)
    s.append((sue.split(' ')[1], {}))

    for v in vls.split(", "):
        val,amt = v.split(": ")
        s[-1][1][val] = int(amt)

print("Part 1:", part1(s))
print("Part 2:", part2(s))

File: 2015/rust/16.rs
// Advent of Code 2015, day 16
// (c) aichingert

use std::collections::HashMap;

const S: [(&'static str, i32); 10] =[("children", 3), ("cats", 7),("samoyeds", 2), ("pomeranians", 3), 
            ("akitas", 0),("vizslas", 0),("goldfish", 5), ("trees", 3), ("cars", 2), ("perfumes", 1)];

fn part1(aunts: &Vec<HashMap<String, i32>>) -> usize {
    let mut ans = (0usize, 0i32);

    for i in 0..aunts.len() {
        let mut cur = 0i32;
        for key in S.iter() {
            if aunts[i].contains_key(key.0) && aunts[i][key.0] == key.1 {
                cur += 1;
            }
        }

        if ans.1 < cur {
            ans = (i+1, cur);
        }
    }

    ans.0
}

fn part2(aunts: &Vec<HashMap<String, i32>>) -> usize {
    let mut ans = (0usize, 0i32);

    for i in 0..aunts.len() {
        let mut cur = 0i32;
        for key in S.iter() {
            if aunts[i].contains_key(key.0) {
                match key.0 {
                    "cats" | "trees" => if key.1 <= aunts[i][key.0] { cur += 1; }
                    "pomeranians" | "goldfish" => if key.1 >= aunts[i][key.0] { cur += 1; }
                    _  => if key.1 == aunts[i][key.0] { cur += 1; }
                };
            }
        }

        if ans.1 < cur {
            ans = (i+1, cur);
        }
    }

    ans.0
}

fn parse() -> Vec<HashMap<String, i32>> {
    let mut aunts = Vec::new();

    for l in std::fs::read_to_string("../input/16").unwrap().lines() {
        let (_, rhs) = l.split_once(": ").unwrap();
        let vls = rhs.split(", ").collect::<Vec<&str>>();
        let mut values: HashMap<String, i32> = HashMap::new();

        for v in vls {
            let (key, value) = v.split_once(": ").unwrap();
            let value = value.parse::<i32>().unwrap();
            values.insert(key.to_string(), value);
        }
        aunts.push(values);
    }

    aunts
}

fn main() {
    let aunts = parse();

    println!("Part 1: {}", part1(&aunts));
    println!("Part 2: {}", part2(&aunts));
}
